src/Compilers/Helpers/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,182 @@
+        /// The base directory for shadow copies. Each instance of
+        /// <see cref="ShadowCopyAnalyzerAssemblyLoader"/> gets its own
+        /// subdirectory under this directory. This is also the starting point
+        /// for scavenge operations.
+        /// </summary>
+        private readonly string _baseDirectory;
+
+        /// <summary>
+        /// The directory where this instance of <see cref="ShadowCopyAnalyzerAssemblyLoader"/>
+        /// will shadow-copy assemblies.
+        /// </summary>
+        private string _shadowCopyDirectory;
+        private Mutex _shadowCopyDirectoryMutex;
+
+        /// <summary>
+        /// Used to generate unique names for per-assembly directories.
+        /// </summary>
+        private int _assemblyDirectoryId = 0;
+
+        public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory = null)
+        {
+            if (baseDirectory != null)
+            {
+                _baseDirectory = baseDirectory;
+            }
+            else
+            {
+                _baseDirectory = Path.Combine(Path.GetTempPath(), "CodeAnalysis", "AnalyzerShadowCopies");
+            }
+
+            Task.Run((Action)DeleteLeftoverDirectories);
+        }
+
+        private void DeleteLeftoverDirectories()
+        {
+            foreach (var subDirectory in Directory.EnumerateDirectories(_baseDirectory))
+            {
+                string name = Path.GetFileName(subDirectory).ToLowerInvariant();
+                Mutex mutex = null;
+                try
+                {
+                    // We only want to try deleting the directory if no one else is currently
+                    // using it. That is, if there is no corresponding mutex.
+                    if (!Mutex.TryOpenExisting(name, out mutex))
+                    {
+                        ClearReadOnlyFlagOnFiles(subDirectory);
+                        Directory.Delete(subDirectory, recursive: true);
+                    }
+                }
+                catch
+                {
+                    // If something goes wrong we will leave it to the next run to clean up.
+                    // Just swallow the exception and move on.
+                }
+                finally
+                {
+                    if (mutex != null)
+                    {
+                        mutex.Dispose();
+                    }
+                }
+            }
+        }
+
+        protected override Assembly LoadCore(string fullPath)
+        {
+            if (_shadowCopyDirectory == null)
+            {
+                _shadowCopyDirectory = CreateUniqueDirectoryForProcess();
+            }
+
+            string assemblyDirectory = CreateUniqueDirectoryForAssembly();
+            string shadowCopyPath = CopyFileAndResources(fullPath, assemblyDirectory);
+
+            return Assembly.LoadFrom(shadowCopyPath);
+        }
+
+        private string CopyFileAndResources(string fullPath, string assemblyDirectory)
```

Comment:
The `InMemoryAssemblyLoader` that we're replacing doesn't handle exceptions, instead leaving it to the caller. I've maintained that design here.

However, I'll make this clear in the doc comments, and I'll double check that the `AppDomain.AssemblyResolve` event handle handler doesn't leak exceptions.